File: golang/monothread_sort/quick_sort.go
package monothread_sort

type quickSort struct{}

func NewQuickSort() *quickSort {
	return &quickSort{}
}

func (ss *quickSort) Sort(array []int) {
	ss.quickSort(array, 0, len(array)-1)
}

func (ss *quickSort) quickSort(array []int, low int, high int) {
	if low < high {
		pi := ss.partition(array, low, high)

		ss.quickSort(array, low, pi-1)
		ss.quickSort(array, pi+1, high)
	}
}

func (ss *quickSort) partition(array []int, low int, high int) int {
	pivot := array[high]

	i := (low - 1)

	for j := low; j <= high-1; j++ {
		if array[j] < pivot {
			i++
			array[i], array[j] = array[j], array[i]
		}
	}

	array[i+1], array[high] = array[high], array[i+1]

	return i + 1
}

File: cpp/monothread_sort/quick_sort.cpp
namespace quick_sort
{
  using namespace std;

  int partition(vector<int> *array, int low, int high)
  {
    int pivot = array->operator[](high);

    int i = low - 1;

    for (int j = low; j <= high - 1; j++)
    {
      if (array->operator[](j) < pivot)
      {
        i++;
        int aux = array->operator[](i);
        array->operator[](i) = array->operator[](j);
        array->operator[](j) = aux;
      }
    }

    int aux = array->operator[](i + 1);
    array->operator[](i + 1) = array->operator[](high);
    array->operator[](high) = aux;

    return i + 1;
  }

  void quickSort(vector<int> *array, int low, int high)
  {
    if (low < high)
    {
      int pi = partition(array, low, high);

      quickSort(array, low, pi - 1);
      quickSort(array, pi + 1, high);
    }
  }

  void sort(vector<int> *array)
  {
    quickSort(array, 0, array->size() - 1);
  }
}
